### PR TITLE
twitch-cli: init at 1.1.19

### PR DESCRIPTION
**Pull request by BenediktBroich**
###### Description of changes

Initial commit for [twitch cli](https://dev.twitch.tv/docs/cli/).

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes (or backporting 23.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/twitch-cli/default.nix [hunk 1]
```diff
@@ -0,0 +1,28 @@
+{ lib
+, buildGoModule
+, fetchFromGitHub
+}:
+
+buildGoModule rec {
+  pname = "twitch";
```
Comment (Kranzes) [inline pkgs/development/tools/twitch-cli/default.nix:7]:
```suggestion
  pname = "twitch-cli";
```

File: pkgs/development/tools/twitch-cli/default.nix [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+  src = fetchFromGitHub {
+    owner = "twitchdev";
+    repo = "twitch-cli";
+    rev = "v${version}";
+    hash = "sha256-osR7+oQ0QBIk/OrMb4Txhb31zELi+eoM47GsWgF/PAg=";
+  };
+
+  vendorHash = "sha256-OhcRMXY8s+XciF+gV3cZ8fxtzo9+I76tBPZ0xG8ddHU=";
+
+  doCheck = false;
```
Comment (Kranzes) [inline pkgs/development/tools/twitch-cli/default.nix:19]: Add a comment on why the checks are disabled.

Comment (BenediktBroich) [reply pkgs/development/tools/twitch-cli/default.nix:19]:
```
=== RUN   TestApiRequest
    test_setup.go:17: 
                Error Trace:    test_setup.go:17
                                                        api_request_test.go:15
                Error:          Expected nil, but got: &fs.PathError{Op:"mkdir", Path:"/homeless-shelter", Err:0xd}
                Test:           TestApiRequest
Error reading body: Get "potato": unsupported protocol scheme ""--- FAIL: TestApiRequest (0.00s)
=== RUN   TestNewRequest
    test_setup.go:17: 
                Error Trace:    test_setup.go:17
                                                        api_test.go:23
                Error:          Expected nil, but got: &fs.PathError{Op:"mkdir", Path:"/homeless-shelter", Err:0xd}
                Test:           TestNewRequest
{
  "data": []
}
{"data":[]}
{"data":[],"pagination":{"cursor":"test"}}
Endpoint responded with status 204
{"data":[]}
--- FAIL: TestNewRequest (0.00s)
=== RUN   TestValidOptions
    test_setup.go:17: 
                Error Trace:    test_setup.go:17
                                                        api_test.go:73
                Error:          Expected nil, but got: &fs.PathError{Op:"mkdir", Path:"/homeless-shelter", Err:0xd}
                Test:           TestValidOptions
--- FAIL: TestValidOptions (0.00s)
=== RUN   TestGetClientInformation
    test_setup.go:17: 
                Error Trace:    test_setup.go:17
                                                        api_test.go:83
                Error:          Expected nil, but got: &fs.PathError{Op:"mkdir", Path:"/homeless-shelter", Err:0xd}
                Test:           TestGetClientInformation
--- FAIL: TestGetClientInformation (0.00s)
=== RUN   TestNewClient
    test_setup.go:17: 
                Error Trace:    test_setup.go:17
                                                        client_test.go:17
                Error:          Expected nil, but got: &fs.PathError{Op:"mkdir", Path:"/homeless-shelter", Err:0xd}
                Test:           TestNewClient
--- FAIL: TestNewClient (0.00s)
```

File: pkgs/development/tools/twitch-cli/default.nix (hunk 1 above)
Comment (BenediktBroich) [reply pkgs/development/tools/twitch-cli/default.nix:7]: When you type twitch -h you get the help/documentation: `Use "twitch [command] --help" for more information about a command.`. So i named it twitch on purpose.

Comment (Kranzes) [reply pkgs/development/tools/twitch-cli/default.nix:7]: set meta.mainProgram = "twitch" and pname to "twitch-cli"

File: pkgs/development/tools/twitch-cli/default.nix (hunk 2 above)
Comment (Kranzes) [reply pkgs/development/tools/twitch-cli/default.nix:19]: Probably requires networking then, comment that.

**Comment by BenediktBroich** [comment]
@ofborg build twitch-cli

**Comment by BenediktBroich** [comment]
@ofborg build twitch-cli

File: pkgs/development/tools/twitch-cli/default.nix [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+  src = fetchFromGitHub {
+    owner = "twitchdev";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-osR7+oQ0QBIk/OrMb4Txhb31zELi+eoM47GsWgF/PAg=";
+  };
+
+  vendorHash = "sha256-OhcRMXY8s+XciF+gV3cZ8fxtzo9+I76tBPZ0xG8ddHU=";
+
+  doCheck = false; # fails
```
Comment (figsoda) [inline pkgs/development/tools/twitch-cli/default.nix:19]:
```suggestion
  preCheck = ''
    export HOME=$(mktemp -d)
  '';

  __darwinAllowLocalNetworking = true;
```

File: pkgs/development/tools/twitch-cli/default.nix [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+  meta = with lib; {
+    description = "The official Twitch CLI to make developing on Twitch easier";
+    homepage = "https://github.com/twitchdev/twitch-cli";
+    license = licenses.asl20;
+    mainProgram = "twitch";
+    maintainers = with maintainers; [ benediktbroich ];
+    platforms = platforms.all;
```
Comment (figsoda) [inline pkgs/development/tools/twitch-cli/default.nix:27]: we can use `buildGoModule`'s defaults here

```suggestion
```

File: pkgs/development/tools/twitch-cli/default.nix [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+  src = fetchFromGitHub {
+    owner = "twitchdev";
+    repo = pname;
+    rev = "v${version}";
+    hash = "sha256-osR7+oQ0QBIk/OrMb4Txhb31zELi+eoM47GsWgF/PAg=";
+  };
+
+  vendorHash = "sha256-OhcRMXY8s+XciF+gV3cZ8fxtzo9+I76tBPZ0xG8ddHU=";
+
```
Comment (figsoda) [inline pkgs/development/tools/twitch-cli/default.nix:18]: this fixes `twitch-cli version`

```suggestion

  ldflags = [
    "-s"
    "-w"
    "-X=main.buildVersion=${version}"
  ];

```

File: pkgs/development/tools/twitch-cli/default.nix [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+  doCheck = false; # fails
+
+  meta = with lib; {
+    description = "The official Twitch CLI to make developing on Twitch easier";
+    homepage = "https://github.com/twitchdev/twitch-cli";
+    license = licenses.asl20;
+    mainProgram = "twitch";
```
Comment (figsoda) [inline pkgs/development/tools/twitch-cli/default.nix:25]: `twitch-cli` is the only binary I got

```suggestion
```

**Comment by BenediktBroich** [comment]
@ofborg build twitch-cli

**Comment by BenediktBroich** [comment]
@ofborg build twitch-cli

**Review by figsoda — APPROVED** [review]

**Pull request merged**